### PR TITLE
Fix issue 22863 - -main doesn't work anymore when used for linking only

### PR DESCRIPTION
**Pull request by dkorpel**
Porting https://github.com/ldc-developers/ldc/pull/3938 to DMD

### CODE REVIEW

File: src/dmd/mars.d [hunk 1]
```diff
@@ -574,7 +574,15 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
...
     if (params.addMain && !global.hasMainFunction)
-        modules.push(moduleWithEmptyMain());
+    {
+        auto mainModule = moduleWithEmptyMain();
+        modules.push(mainModule);
+
+        if (!params.oneobj)
+            params.objfiles.push(mainModule.objfile.toChars());
+        else if (modules.length == 1)
+            params.objfiles.insert(0, mainModule.objfile.toChars());
```
Comment (kinke) [inline src/dmd/mars.d:584]: Inserting at position 0 is just an LDC specificum for `-singleobj`, DMD shouldn't need it and probably just append it, as for the `!oneobj` case.

Comment (ljmf00) [reply src/dmd/mars.d]: There are some `.debug_ranges` conflicts on debug info for `-main` generated functions. Perhaps, is this somewhat related?

**Comment by dkorpel** [comment]
circleci:
```
ERROR: Found the following residual temporary files.
ERROR: Temporary files should be stored in `test_results` or explicitly removed.
?? test/foo
```
Not sure why it's not deleted, dmd gets passed `-od=${OUTPUT_BASE}` and `rm_retry -r ${OUTPUT_BASE}` follows.

**Comment by kinke** [comment]
`-od` just specifies the object files dir IIRC, doesn't affect the executable output dir.

**Comment by WalterBright** [comment]
I'll defer to @kinke on this one. -Walter

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**